src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs

```diff
@@ -1364,38 +1358,37 @@ internal static unsafe string UInt32ToDecStr(uint value)
             }
 
             string result = string.FastAllocateString(bufferLength);
-            fixed (char* buffer = result)
+            Span<char> buffer = new Span<char>(ref result.GetRawStringData(), bufferLength);
+
+            int p = bufferLength;
+            do
             {
-                char* p = buffer + bufferLength;
-                do
-                {
-                    uint remainder;
-                    (value, remainder) = Math.DivRem(value, 10);
-                    *(--p) = (char)(remainder + '0');
-                }
-                while (value != 0);
-                Debug.Assert(p == buffer);
+                uint remainder;
+                (value, remainder) = Math.DivRem(value, 10);
+                buffer[--p] = (char)(remainder + '0');
```

Comment:
Here's another place where bounds checks are being inserted

Reply:
Can't find a way to remove bound checks, specially because we need to calculate the remainder from the end and JIT doesn't optimize reverse loops